OpenRA.Mods.Common/AI/BaseBuilder.cs

```diff
@@ -199,24 +266,34 @@ ActorInfo ChooseBuildingToBuild(ProductionQueue queue)
 				if (ai.Info.BuildingLimits.ContainsKey(name) && ai.Info.BuildingLimits[name] <= count)
 					continue;
 
+				if (!waterAvailable && ai.Info.BuildingCommonNames.ContainsKey("NavalProduction")
+					&& ai.Info.BuildingCommonNames["NavalProduction"].Contains(name))
+					continue;
+
+				if (waterAvailable && ai.Info.BuildingCommonNames.ContainsKey("NavalProduction")
```

Comment:
This is mostly duplicating the check from above - how about pulling it out into a temporary variable above?